File: IOS/ChordRecognizeGenerate/ChordRecognizeGenerate/Triad.swift
//
//  Triad.swift
//  ChordRecognizeGenerate
//
//  Created by Derrick Ward on 1/2/20.
//  Copyright © 2020 Derrick Ward. All rights reserved.
//

import Foundation

@objc public class Triad : NSObject, NSSecureCoding
{
    enum Required
    {
        case Third, Fifth
    }
    
    public static var supportsSecureCoding = true
    public var third : NoteRootOffset
    public var fifth : NoteRootOffset
    public var quality : Quality
    public var requiredNote : NoteRootOffset
    
    override init()
    {
        third = 3
        fifth = 5
        quality = Quality.Maj
        requiredNote = 3
        super.init()
    }
    
    init(third : NoteRootOffset, fifth : NoteRootOffset, required : NoteRootOffset,quality : Quality)
    {
        self.third = third
        self.fifth = fifth
        self.quality = quality
        self.requiredNote = required
    }
    
    public required init?(coder: NSCoder) {
       
        third = NoteRootOffset(coder.decodeInteger(forKey: "third"))
        fifth = NoteRootOffset(coder.decodeInteger(forKey: "fifth"))
        quality = Quality(rawValue: Int(coder.decodeInt32(forKey: "quality")))!
        requiredNote = NoteRootOffset(coder.decodeInteger(forKey: "requiredNote"))
        super.init()
    }
    
    public func encode(with coder: NSCoder) {
        coder.encode(Int(third), forKey: "third")
        coder.encode(Int(fifth), forKey: "fifth")
        coder.encode(quality.rawValue, forKey: "quality")
        coder.encode(Int(requiredNote), forKey: "requiredNote")
    }
    
    func addNotes(chordNotes : inout [NoteRootOffset], inversion : Inversion)
    {
        var notes = [third,fifth]
        Triad.invert(notes: &notes, inversion: inversion)
        chordNotes.append(contentsOf: notes)
    }
    
    func isSuspended() -> Bool
    {
        return quality == Quality.sus2 || quality == Quality.sus4
    }
    
    func matchesNote(note : ChordNote) -> Bool
    {
        return self.third == note || self.fifth == note
    }
    
    static func extractTriadNotes(rootIdx: Int, notes : [ChordNote]) -> [NoteRootOffset : Int]
    {
        var triadNotes = [NoteRootOffset : Int]()
        let baseNote = notes[rootIdx]
        for i in 0..<notes.count
        {
            let noteOffset = ChordDefs.offsetInChord(rootNote: baseNote, note: notes[i])
            
            if noteOffset == 0 && triadNotes[0] == nil
            {
                triadNotes[0] = i
                continue
            }
            
            if noteOffset >= 2 && noteOffset <= 8
            {
                triadNotes[noteOffset] = i
            }
        }
        
        return triadNotes
    }
    
    func matches(notes : [ChordNote], triadNotes: [NoteRootOffset : Int]) -> TriadMatch
    {
        let result = TriadMatch()
        result.triadType = self.quality
        result.triad = self
        result.rootIdx = triadNotes[0]!
        
        let rootNote = notes[triadNotes[0]!]
        
        if let _ = triadNotes[requiredNote]
        {
            if let fifthNoteIdx = triadNotes[fifth]
            {
                let fifthNote = notes[fifthNoteIdx]
                addNoteToTriadMatch(rootNote: rootNote, note: fifthNote, offset: fifth, result: result)
            }
            else
            {
                if let sharpFifthIdx = triadNotes[ChordDefs.sharpFifth]
                {
                    let fifthNote = notes[sharpFifthIdx]
                    result.alteredFifth = fifthNote
                    result.alteredFifthSign = Sign.sharp
                    
                    addNoteToTriadMatch(rootNote: rootNote, note: fifthNote, offset: ChordDefs.sharpFifth, result: result)
                }
                else if let flatFifthIdx = triadNotes[ChordDefs.flatFifth]
                {
                     let fifthNote = notes[flatFifthIdx]
                     result.alteredFifth = fifthNote
                     result.alteredFifthSign = Sign.flat
                      addNoteToTriadMatch(rootNote: rootNote, note: fifthNote, offset: ChordDefs.flatFifth, result: result)
                }
                else
                {
                    result.omission = Degree.five
                }
            }
            
            if let thirdNoteIdx = triadNotes[third]
            {
                let thirdNote = notes[thirdNoteIdx]
                addNoteToTriadMatch(rootNote: rootNote, note: thirdNote, offset: third, result: result)
            }
            else
            {
                result.omission = Degree.three
            }
            
        }
        
        for note in notes
        {
            if !result.triadNotes.contains(note) && abs(Int8(note) - Int8(rootNote)) % 12 != 0
            {
                result.nonTriadNotes.append(note)
            }
        }
        
        return result
    }
    
    func addNoteToTriadMatch(rootNote: ChordNote, note: ChordNote, offset : NoteRootOffset,result : TriadMatch)
    {
        result.triadNotes.append(note)
        
        if note < rootNote
        {
            result.invertedNotes.append(note)
            result.inversion = offset == third ? Inversion.first : Inversion.second
        }
    }
    
    private func isAlternateFifth(offset: NoteRootOffset) -> Bool
    {
        return offset == ChordDefs.flatFifth || offset == ChordDefs.sharpFifth
    }
    
    public static func invert(notes: inout [NoteRootOffset], inversion: Inversion)
    {
            switch inversion {
            case .first:
                notes[0] -= 12
            case .second:
                notes[1] -= 12
            default:
                print("default")
            }
    }
}

public class TriadMatch
{
    var nonTriadNotes = [ChordNote]()
    var invertedNotes = [ChordNote]()
    var omission : Degree?
    var triadNotes = [ChordNote]()
    weak var triad : Triad!
    var rootIdx : Int = 0
    var triadType = Quality.Maj
    var alteredFifth : ChordNote?
    var alteredFifthSign : Sign = Sign.natural
    var inversion = Inversion.none
    public var isHalfDiminished = false
    
    func getScore() -> Int
    {
        if triadNotes.count == 0
        {
            return 0
        }
        else
        {
            let score = triadNotes.count - rootIdx
                - (alteredFifth == nil ? 0 : 1)
        
            return score
        }
        
    }
}

File: Android/ChordRecGenLib/src/main/java/com/blindintuition/chordrecgenlib/Triad.kt
package com.blindintuition.chordrecgenlib

import blindintuition.midi3d.midi.MIDIConstants
import java.lang.Math.abs
import java.util.*
import kotlin.collections.HashMap

class Triad {

    enum class Required
    {
        Third, Fifth
    }

    var root : NoteRootOffset = 0
    var third : NoteRootOffset
    var fifth : NoteRootOffset
    var quality : Quality
    var requiredNote : NoteRootOffset

    constructor()
    {
        third = 3
        fifth = 5
        quality = Quality.Maj
        requiredNote = 3
    }

    constructor(third : NoteRootOffset, fifth : NoteRootOffset, required : NoteRootOffset,quality : Quality)
    {
        this.third = third
        this.fifth = fifth
        this.quality = quality
        this.requiredNote = required
    }

    fun addNotes(chordNotes : ArrayList<NoteRootOffset>, inversion : Inversion)
    {
        val notes = arrayOf(root,third,fifth)
        invert(notes, inversion)
        chordNotes.addAll(notes)
    }

    fun isSuspended() : Boolean
    {
        return quality == Quality.sus2 || quality == Quality.sus4
    }

    fun matchesNote(note : ChordNote) : Boolean
    {
        return third == note || fifth == note
    }

    fun matches(notes : Array<ChordNote>, triadNotes: Map<NoteRootOffset,Int>) : TriadMatch
    {
        val result = TriadMatch()
        result.triadType = quality
        result.triad = this
        result.rootIdx = triadNotes[0]!!
        result.notes = notes

        val rootNote = notes[triadNotes[0]!!]

        if (triadNotes[requiredNote] != null)
        {
            triadNotes[fifth]?.also { fifthNoteIdx ->
                val fifthNote = notes[fifthNoteIdx]
                addNoteToTriadMatch(rootNote, fifthNote, fifth, result)
            } ?:
            run {
                triadNotes[ChordDefs.sharpFifth]?.also { sharpFifthIdx ->
                    val fifthNote = notes[sharpFifthIdx]
                    result.alteredFifth = fifthNote
                    result.alteredFifthSign = Sign.sharp
                    addNoteToTriadMatch( rootNote, fifthNote, ChordDefs.sharpFifth, result)
                } ?:
                triadNotes[ChordDefs.flatFifth]?.also { flatFifthIdx ->
                    val fifthNote = notes[flatFifthIdx]
                    result.alteredFifth = fifthNote
                    result.alteredFifthSign = Sign.flat
                    addNoteToTriadMatch(rootNote,fifthNote, ChordDefs.flatFifth, result)
                } ?:
                run {
                    result.omission = Degree.five
                }
            }

            triadNotes[third]?.also { thirdNoteIdx ->
                val thirdNote = notes[thirdNoteIdx]
                addNoteToTriadMatch(rootNote, thirdNote,third,result)
            } ?:
            run {
                result.omission = Degree.three
            }
        }

        for (note in notes)
        {
            if (!result.triadNotes.contains(note) && abs(note - rootNote) % 12 != 0)
            {
                result.nonTriadNotes.add(note)
            }
        }

        return result
    }


    fun addNoteToTriadMatch(rootNote: ChordNote, note: ChordNote, offset : NoteRootOffset,result : TriadMatch)
    {
        result.triadNotes.add(note)

        if (note < rootNote)
        {
            result.invertedNotes.add(note)
            result.inversion = if(offset == third) Inversion.first else Inversion.second
        }
    }

    companion object Triad {

        fun extractTriadNotes(rootIdx: Int, notes : Array<ChordNote>) : HashMap<NoteRootOffset,Int>
        {
            val triadNotes = hashMapOf<NoteRootOffset,Int>()
            val baseNote = notes[rootIdx]
            for (i in 0 until notes.size)
            {
                val noteOffset = ChordDefs.offsetInChord(baseNote,notes[i])

                if ((noteOffset == 0.toByte()) && (triadNotes[0] == null))
                {
                    triadNotes[0] = i
                    continue
                }

                if (noteOffset >= 2 && noteOffset <= 8)
                {
                    triadNotes[noteOffset] = i
                }
            }

            return triadNotes
        }

        private fun isAlternateFifth(offset: NoteRootOffset): Boolean {
            return offset == ChordDefs.flatFifth || offset == ChordDefs.sharpFifth
        }

        fun invert(notes: Array<NoteRootOffset>, inversion: Inversion) {
            when(inversion)
            {
                Inversion.first -> {
                    notes[0] = (notes[0] + 12).toNoteRootOffset()
                }
                Inversion.second -> {
                    notes[0] = (notes[0] + 12).toNoteRootOffset()
                    notes[1] = (notes[1] + 12).toNoteRootOffset()
                }
                Inversion.third -> {
                    notes[0] = (notes[0] + 12).toNoteRootOffset()
                    notes[1] = (notes[1] + 12).toNoteRootOffset()
                    notes[2] = (notes[2] + 12).toNoteRootOffset()
                }
                else -> {

                }
            }
        }
    }

    class TriadMatch
    {
        var nonTriadNotes = arrayListOf<ChordNote>()
        var invertedNotes = arrayListOf<ChordNote>()
        lateinit var notes : Array<ChordNote>
        var omission : Degree? = null
        var triadNotes = arrayListOf<ChordNote>()
        lateinit var triad : com.blindintuition.chordrecgenlib.Triad
        var rootIdx : Int = 0
        var triadType = Quality.Maj
        var alteredFifth : ChordNote? = null
        var alteredFifthSign : Sign = Sign.natural
        var inversion = Inversion.none
        var isHalfDiminished = false

        fun getName() : String
        {
            val rootNote = notes[rootIdx]
            val rootName = MIDIConstants.stringForNote(rootNote)
            val alteration = if(alteredFifth == null) "" else String.format("%s%s",Degree.five.text,alteredFifthSign.text)
            return String.format("%s%s %s",rootName,triadType.text,alteration)
        }

        fun getScore() : Int
        {
            if(triadNotes.size == 0)
            {
                return 0
            }
            else
            {
                val score = triadNotes.size - rootIdx
                    + (if(omission == null) 4 else 0)
                    + (if(alteredFifth == null) 4 else 0)

                return score
            }
        }
    }
}

